enterprise/server/cmd/goinit/main.go

```diff
@@ -494,6 +526,125 @@ func runVMExecServer(ctx context.Context) error {
+}
+
+func runVMDNSServer(ctx context.Context) error {
+	log.Infof("Starting vm dns server on port 53")
+
+	var dnsOverrides []*networking.DnsOverride
+	dnsOverridesJSON, err := fetchMMDSKey("dns_overrides")
+	if err != nil {
+		return err
+	}
+	if err := json.Unmarshal(dnsOverridesJSON, &dnsOverrides); err != nil {
+		return err
+	}
+
+	mux := dns.NewServeMux()
+
+	// Ensure hostnames end with '.' so they are not resolved as relative names.
+	for _, o := range dnsOverrides {
+		if o.RedirectToHostname == "" {
+			return status.InvalidArgumentError("empty redirect hostname")
+		}
+		if !strings.HasSuffix(o.HostnameToOverride, ".") {
+			o.HostnameToOverride += "."
+		}
+	}
+
+	mux.Handle(".", &vmDNSServer{overrides: dnsOverrides})
+	s := &dns.Server{
+		Addr:    "127.0.0.1:53",
+		Net:     "udp4",
+		Handler: mux,
+	}
+	return s.ListenAndServe()
+}
+
+func (s *vmDNSServer) ServeDNS(w dns.ResponseWriter, req *dns.Msg) {
+	for _, o := range s.overrides {
+		if strings.HasSuffix(req.Question[0].Name, o.HostnameToOverride) && !strings.HasSuffix(req.Question[0].Name, o.RedirectToHostname) {
+			s.overrideResponse(w, req, o.RedirectToHostname)
+			return
+		}
+	}
+
+	// For all other requests, forward to a remote DNS server.
+	rsp, err := s.forwardToRemoteServer(req)
+	if err != nil {
+		s.returnFailureMsg(w, req)
+		return
+	}
+
+	if err := w.WriteMsg(rsp); err != nil {
+		log.Errorf("could not send dns reply: %s", err)
+	}
+}
+
+// overrideResponse returns a DNS response pointing requests originally directed
+// to `hostname_to_override` => `redirect_to_hostname`.
+// The response also contains the A record for `redirect_to_hostname` so the
+// client can follow the request to the final IP addr it should hit.
+func (s *vmDNSServer) overrideResponse(w dns.ResponseWriter, req *dns.Msg, redirectToHostname string) {
+	rspMsg := &dns.Msg{}
+	rspMsg.SetReply(req)
+	rspMsg.Rcode = dns.RcodeSuccess
+	rspMsg.RecursionAvailable = true
+
+	cname := &dns.CNAME{
+		Hdr: dns.RR_Header{
+			Name:   req.Question[0].Name,
+			Rrtype: dns.TypeCNAME,
+			Class:  dns.ClassINET,
+			Ttl:    60,
+		},
+		Target: redirectToHostname,
+	}
+
+	// Get A record for `redirectToHostname` from remote DNS server.
+	forwardMsg := &dns.Msg{}
+	forwardMsg.SetQuestion(redirectToHostname, dns.TypeA)
+	forwardMsg.RecursionDesired = true
+	rsp, err := s.forwardToRemoteServer(forwardMsg)
+	if err != nil {
+		s.returnFailureMsg(w, req)
+		return
+	}
+
+	rspMsg.Answer = append(rspMsg.Answer, cname)
+	rspMsg.Answer = append(rspMsg.Answer, rsp.Answer...)
+
+	if err := w.WriteMsg(rspMsg); err != nil {
+		log.Errorf("could not send dns reply: %s", err)
+	}
+}
+
+// forwardToRemoteServer forwards a DNS query to remote DNS servers. Other than configured
+// queries that should be overwritten by the local DNS server, most should be
+// forwarded on.
+func (s *vmDNSServer) forwardToRemoteServer(req *dns.Msg) (*dns.Msg, error) {
+	upstreams := []string{"8.8.8.8:53", "8.8.4.4:53", "1.1.1.1:53"}
```

Comment:
do we need this fallback logic since these IPs are already listed in /etc/hosts? or can we just return a not found response (idk what the equivalent is of that in the DNS protocol)

Reply:
The DNS fallbacks in /etc/resolv.conf will only be used if the first DNS entry is completely unresponsive. If the first DNS server returned a not found error, the error would be immediately returned. So unfortunately this local DNS server has to explicitly forward on queries it can't resolve itself